src/main/scala/vee/db/Caches.scala

```diff
@@ -81,125 +52,108 @@ trait Caches extends Blockchain {
-                         assetBalances: Map[BigInt, Map[ByteStr, Long]],
-                         leaseBalances: Map[BigInt, LeaseBalance],
+                         veeBalances: Map[BigInt, Long],
+                         leaseBalances: Map[BigInt, LeaseInfo],
                          leaseStates: Map[ByteStr, Boolean],
-                         transactions: Map[ByteStr, (Transaction, Set[BigInt])],
-                         addressTransactions: Map[BigInt, List[(Int, ByteStr)]],
-                         reissuedAssets: Map[ByteStr, AssetInfo],
-                         filledQuantity: Map[ByteStr, VolumeAndFee],
-                         scripts: Map[BigInt, Option[Script]],
-                         data: Map[BigInt, AccountDataInfo],
-                         aliases: Map[Alias, BigInt],
-                         sponsorship: Map[AssetId, Sponsorship]): Unit
-
-  override def append(diff: Diff, carryFee: Long, block: Block): Unit = {
-    heightCache += 1
-    scoreCache += block.blockScore()
-    lastBlockCache = Some(block)
-
-    val newAddresses = Set.newBuilder[Address]
-    newAddresses ++= diff.portfolios.keys.filter(addressIdCache.get(_).isEmpty)
-    for ((_, _, addresses) <- diff.transactions.values; address <- addresses if addressIdCache.get(address).isEmpty) {
-      newAddresses += address
-    }
-
-    val newAddressIds = (for {
-      (address, offset) <- newAddresses.result().zipWithIndex
-    } yield address -> (lastAddressId + offset + 1)).toMap
-
-    def addressId(address: Address): BigInt = (newAddressIds.get(address) orElse addressIdCache.get(address)).get
-
-    lastAddressId += newAddressIds.size
-
-    val wavesBalances = Map.newBuilder[BigInt, Long]
-    val assetBalances = Map.newBuilder[BigInt, Map[ByteStr, Long]]
-    val leaseBalances = Map.newBuilder[BigInt, LeaseBalance]
-    val newPortfolios = Map.newBuilder[Address, Portfolio]
-
-    for ((address, portfolioDiff) <- diff.portfolios) {
-      val newPortfolio = portfolioCache.get(address).combine(portfolioDiff)
-      if (portfolioDiff.balance != 0) {
-        wavesBalances += addressId(address) -> newPortfolio.balance
-      }
-
-      if (portfolioDiff.lease != LeaseBalance.empty) {
-        leaseBalances += addressId(address) -> newPortfolio.lease
-      }
-
-      if (portfolioDiff.assets.nonEmpty) {
-        val newAssetBalances = for { (k, v) <- portfolioDiff.assets if v != 0 } yield k -> newPortfolio.assets(k)
-        if (newAssetBalances.nonEmpty) {
-          assetBalances += addressId(address) -> newAssetBalances
-        }
+                         transactions: Map[ByteStr, (ProcessedTransaction, Set[BigInt])],
+                         addressTransactions: Map[BigInt, List[ByteStr]]
+                        ): Unit
+
+  override def appendBlock(block: Block)(consensusValidation: => Either[ValidationError, BlockDiff]): Either[ValidationError, BlockDiff] = {
+    if ((heightCache == 0) || (this.lastBlock.get.uniqueId == block.reference)) {
```

Comment:
does it batter to use 
```scala
if ((height() == 0) || (this.lastBlock.get.uniqueId == block.reference)) consensusValidation.map { blockDiff =>
...
blockDiff
}
else {
      Left(GenericError(s"Parent ${block.reference} of block ${block.uniqueId} does not match last local block ${this.lastBlock.map(_.uniqueId)}"))
}
// follow the HistoryWriterImpl.scala
```
since `consensusValidation match` return `None` if `consensusValidation` `Left`

Reply:
`Either is right-biased, which means that Right is assumed to be the default case to operate on. If it is Left, operations like map, flatMap, ... return the Left value unchanged`
[https://www.scala-lang.org/api/2.12.0/scala/util/Either.html](url)